uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
+            assigned a badge number. If neither parameter is given, always returns False.
 
-            if badge_type not in c.PREASSIGNED_BADGE_TYPES:
-                return 0
+            :param attendee: Passing an existing attendee allows us to check for a new badge num whenever the attendee
+            is updated, particularly for when they are checked in.
+            :param badge_type: Allows checking for a new badge number before adding/updating the Attendee() object.
+            :return:
+            """
+            if not badge_type and attendee:
+                badge_type = attendee.badge_type
+            elif not badge_type and not attendee:
+                return None
 
-            sametype = self.query(Attendee).filter(Attendee.badge_type == badge_type,
-                                                   Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
-                                                   Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
-            if sametype.count():
-                next = sametype.order_by(Attendee.badge_num.desc()).first().badge_num
-                if old_badge_num and next == old_badge_num:
-                    next = next  # Prevents incrementing if the current badge already has the highest badge number in the range.
+            if c.NUMBERED_BADGES:
+                if attendee:
+                    return (badge_type in c.PREASSIGNED_BADGE_TYPES or attendee.checked_in)\
+                           and attendee.paid != c.NOT_PAID and not attendee.is_unassigned and attendee.badge_status != c.INVALID_STATUS
                 else:
-                    next += 1
-            else:
-                next = c.BADGE_RANGES[badge_type][0]
+                    return badge_type in c.PREASSIGNED_BADGE_TYPES
+
+        def get_next_badge_num(self, badge_type):
+            """
+            Returns the next badge available for a given badge type. This is essentially a wrapper for auto_badge_num
+            that accounts for new or changed objects in the session.
 
+            :param badge_type: Used to pass to auto_badge_num and to ignore objects in the session that aren't within
+            the badge type's range.
+            :return:
+            """
+            new_badge_num = self.auto_badge_num(badge_type)
             # Adjusts the badge number based on badges in the session
             for attendee in [m for m in chain(self.new, self.dirty) if isinstance(m, Attendee)]:
-                if attendee.badge_type == badge_type and attendee.badge_num <= c.BADGE_RANGES[badge_type][1]:
-                    next = max(next, 1 + attendee.badge_num)
-
-            return next
+                if attendee.badge_type == badge_type and attendee.badge_num is not None\
+                        and attendee.badge_num <= c.BADGE_RANGES[badge_type][1]\
+                        and attendee.badge_num <= self.auto_badge_num(badge_type):
```

Comment:
This line was added to accommodate our new 'fill-in-the-gap' functionality in `auto_badge_num` - without it, dirty/new sessions would push badge numbers over the gap.